pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionMetaDataTopicTailer.java

```diff
@@ -77,12 +89,25 @@ public void run() {
                     if (!(th instanceof InterruptedException || th.getCause() instanceof InterruptedException)) {
                         log.warn("Encountered error when metadata tailer is not running", th);
                     }
-                    return;
                 }
             }
         }
     }
 
+    public void stopWhenNoMoreMessages() {
+        stopOnNoMessageAvailable = true;
+        readerThread.interrupt();
+        // We need to wait here till the thread exits to make sure that the reader is up to date
+        while (true) {
+            try {
+                readerThread.join();
+                return;
+            } catch (InterruptedException e) {
```

Comment:
we shouldn't need to catch any exceptions.  The thread needs to complete without any errors

Reply:
removed